linera-core/src/chain_worker/state/mod.rs

```diff
@@ -306,39 +306,56 @@ where
     }
 
     /// Returns an error if the block requires a blob we don't have.
-    /// Looks for the blob in: chain manager's pending blobs and storage.
-    async fn check_no_missing_blobs(
+    /// Looks for the blob in: chain manager's proposed blobs and storage.
+    async fn get_proposed_blobs_and_check_storage(
         &self,
-        required_blob_ids: &HashSet<BlobId>,
-    ) -> Result<(), WorkerError> {
-        let pending_blobs = &self.chain.manager.get().pending_blobs;
-        let missing_blob_ids = required_blob_ids
-            .iter()
-            .filter(|blob_id| !pending_blobs.contains_key(blob_id))
-            .cloned()
-            .collect::<Vec<_>>();
-
-        let missing_blob_ids = self
-            .storage
-            .missing_blobs(missing_blob_ids.as_slice())
-            .await?;
+        blob_ids: &HashSet<BlobId>,
+    ) -> Result<Vec<Blob>, WorkerError> {
+        let published_blobs = self.chain.manager.get().proposed_blobs();
+        self.get_blobs_and_check_storage(blob_ids, &published_blobs, &BTreeMap::new())
+            .await
+    }
 
-        if missing_blob_ids.is_empty() {
-            return Ok(());
+    async fn get_blobs_and_check_storage(
+        &self,
+        blob_ids: &HashSet<BlobId>,
+        published_blobs: &BTreeMap<BlobId, Blob>,
+        used_blobs: &BTreeMap<BlobId, Blob>,
+    ) -> Result<Vec<Blob>, WorkerError> {
+        let mut found_blobs = Vec::new();
+        let mut missing_blob_ids = Vec::new();
+        for blob_id in blob_ids {
+            if let Some(blob) = published_blobs.get(blob_id) {
+                found_blobs.push(blob.clone());
+            } else if let Some(blob) = used_blobs.get(blob_id) {
+                found_blobs.push(blob.clone());
+            } else {
+                missing_blob_ids.push(*blob_id);
+            }
         }
+        let not_found_blob_ids = self.storage.missing_blobs(&missing_blob_ids).await?;
```

Comment:
(Just for my own understanding: We don't need to add the ones we _did_ find in storage to the `found_blobs` here because we're never using this in a context where we'd need to add them to the chain manager?)

Reply:
I was under the impression that if something was already confirmed in storage, we didn't need to pass that along because we would just end up trying to write it to storage again needlessly. This was before our talk about the locked blobs 😅
So we actually do need to read from storage here instead of just checking it.